### PR TITLE
feat: boolean calculator 1st attempt

### PR DESCRIPTION
**Pull request by karpov-kir**
This is my first attempt to implement the boolean calculator assignment. The solution supports nested parenthesis, it is also case, and space insensitive.

### CODE REVIEW

File: 1_The_Physical/Guess_Point_Level_1_Stateless/3_Manage_Complexity/3_2_Boolean_Calculator/src/index.ts [hunk 1]
```diff
@@ -10,6 +17,37 @@ export class BooleanCalculator {
       }
     });
 
-    return booleanExpression === 'TRUE';
+    let parsingState = ParsingState.ExpectingBoolean;
```
Comment (stemmlerjs) [inline 1_The_Physical/Guess_Point_Level_1_Stateless/3_Manage_Complexity/3_2_Boolean_Calculator/src/index.ts]: This is the commit right here, where it all kicked off. How much of this did you do upfront? Did you happen to draw this out, do much of it in your mind, or did this just emerge? It's quite the *obvious implementation*.

Categorizing what you're going to expect next as state or logic. Say more on your thinking here?

File: 1_The_Physical/Guess_Point_Level_1_Stateless/3_Manage_Complexity/3_2_Boolean_Calculator/src/index.ts [hunk 2]
```diff
@@ -10,6 +17,37 @@ export class BooleanCalculator {
...
+
+    tokens.forEach((token) => {
+      if (parsingState === ParsingState.ExpectingBoolean) {
+        if (!allowedBooleanTokens.includes(token)) {
+          throw new Error(`Expect a boolean token but go "${token}"`);
+        }
+
+        if (state === undefined) {
+          state = token === 'TRUE';
+        } else {
```
Comment (stemmlerjs) [inline 1_The_Physical/Guess_Point_Level_1_Stateless/3_Manage_Complexity/3_2_Boolean_Calculator/src/index.ts]: Style tip, I always try to remove having to use the else keyword if I can. It just creates a code path that makes things harder to follow, especially if there's another nested conditional.

File: 1_The_Physical/Guess_Point_Level_1_Stateless/3_Manage_Complexity/3_2_Boolean_Calculator/src/index.ts [hunk 3]
```diff
@@ -1,12 +1,23 @@
 const booleanTokens = ['TRUE', 'FALSE'];
 const logicTokens = ['AND', 'OR'];
 const negationToken = 'NOT';
+const openGroupToken = '(';
```
Comment (stemmlerjs) [inline 1_The_Physical/Guess_Point_Level_1_Stateless/3_Manage_Complexity/3_2_Boolean_Calculator/src/index.ts]: Incredible 💥 love to see you pull out the classic data structures for this one. This is originally how I solved this in college. I am enjoying watching you define these concepts of the domain here.

File: 1_The_Physical/Guess_Point_Level_1_Stateless/3_Manage_Complexity/3_2_Boolean_Calculator/src/index.ts (hunk 1 above)
Comment (karpov-kir) [reply 1_The_Physical/Guess_Point_Level_1_Stateless/3_Manage_Complexity/3_2_Boolean_Calculator/src/index.ts]: @stemmlerjs, first of all, I did this assignment at the very begging of the course when we did not have lots of content yet, e.g. I did not know about TPP.

I decided to try only emergent design without the upfront part. This approach was devouring my brain power and sometimes it was quite hard to move forward. But I like the eventual outcome.

If I were to implement the assignment again I would use TPP to complete the assignment in even smaller steps and I would do a little bit of upfront design too.

> Categorizing what you're going to expect next as state or logic. Say more on your thinking here?

This was a part of the emergent design. I just added one more test case and implemented the missing functionality using the obvious implementation approach. It means that once I added `TRUE AND FALSE` it became clear that the logic concept is missing and needed to be implemented so I just added `const allowedLogicTokens = ['AND', 'OR'];` and built an obvious implementation around it.

File: 1_The_Physical/Guess_Point_Level_1_Stateless/3_Manage_Complexity/3_2_Boolean_Calculator/src/index.ts (hunk 2 above)
Comment (karpov-kir) [reply 1_The_Physical/Guess_Point_Level_1_Stateless/3_Manage_Complexity/3_2_Boolean_Calculator/src/index.ts]: True! Reduced complexity in https://github.com/karpov-kir/the-software-essentialist-cource/pull/7/commits/1e8247722ff3c5d2d5c2a7e90d215ca6bcdaf470.

**Comment by karpov-kir** [comment]
I merged the PR since I received and addressed the feedback above, so this iteration of the assignment can be considered as finished.

**Comment by karpov-kir** [comment]
I merged the PR since I received and addressed the feedback above, so this iteration of the assignment can be considered as finished.

**Pull request merged**